client/db/bolt/db.go

```diff
@@ -1624,6 +1624,181 @@ func (idx *timeIndexNewest) add(t uint64, k []byte, b *bbolt.Bucket) {
+	start := time.Now()
+	const batchSize = 1000
+	var (
+		finished   bool
+		olderThanB []byte
+	)
+	if olderThan != nil {
+		olderThanB = uint64Bytes(encode.UnixMilliU(*olderThan))
+	} else {
+		olderThanB = uint64Bytes(encode.UnixMilliU(start))
+	}
+
+	activeMatches := make(map[order.OrderID]struct{})
+	if err := db.View(func(tx *bbolt.Tx) error {
+		// Some archived orders may still be needed for active matches.
+		// Put those order id's in a map to prevent deletion.
+		amb := tx.Bucket(activeMatchesBucket)
+		if err := amb.ForEach(func(k, _ []byte) error {
+			mBkt := amb.Bucket(k)
+			if mBkt == nil {
+				return fmt.Errorf("match %x bucket is not a bucket", k)
+			}
+
+			oidB := mBkt.Get(orderIDKey)
+			var oid order.OrderID
+			copy(oid[:], oidB)
+			activeMatches[oid] = struct{}{}
+			return nil
+		}); err != nil {
+			return fmt.Errorf("unable to get active matches: %v", err)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	nDeletedOrders := 0
+	for !finished {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		nDeletedOrders = 0
+		err := db.Update(func(tx *bbolt.Tx) error {
+			// Run through the archived order bucket storing id's keys for
+			// returning the order data and deletion later.
+			archivedOB := tx.Bucket(archivedOrdersBucket)
+			if archivedOB == nil {
+				return fmt.Errorf("failed to open %s bucket", string(archivedOrdersBucket))
+			}
+			// Retrieve one more than we will delete in order to gauge if
+			// this is the last order to delete.
+			oneOverSize := batchSize + 1
+			filter := func(k []byte, oBkt *bbolt.Bucket) bool {
+				var oid order.OrderID
+				copy(oid[:], k)
+				if order.OrderStatus(intCoder.Uint16(oBkt.Get(statusKey))).IsActive() {
+					db.log.Warnf("active order %v found in inactive bucket", oid)
+					return false
+				}
+				if _, has := activeMatches[oid]; has {
+					return false
+				}
+				timeB := oBkt.Get(updateTimeKey)
+				return bytes.Compare(timeB, olderThanB) <= 0
+			}
+			trios := newestBuckets([]*bbolt.Bucket{archivedOB}, oneOverSize, updateTimeKey, filter)
+
+			// Ignore the last order if it exists. Otherwise there are no
+			// more orders to delete.
+			if len(trios) == oneOverSize {
+				trios = trios[:batchSize]
+			} else {
+				finished = true
+			}
+			for _, trio := range trios {
+				o, err := decodeOrderBucket(trio.k, trio.b)
+				if err != nil {
+					return fmt.Errorf("failed to decode order bucket: %v", err)
+				}
+				if err := archivedOB.DeleteBucket(trio.k); err != nil {
+					return fmt.Errorf("failed to delete order bucket: %v", err)
+				}
+				if perOrderFn != nil {
+					if err := perOrderFn(o); err != nil {
+						return fmt.Errorf("problem performing batch function: %v", err)
+					}
+				}
+				nDeletedOrders++
+			}
+			db.log.Infof("Deleted %d orders from the database in %v.", nDeletedOrders, time.Since(start))
```

Comment:
This PR tests quite well, but one thing that stood out is these logs that show an increasing duration, but always 1000 orders. The log would make sense if it either showed cumulative deleted orders (not just the batch), or it showed the time for the last batch (not the cumulative time).

Reply:
Ah. Will change. It is still working with the assumption that the caller is doing this method once at a time.

Reply:
Ok https://github.com/decred/dcrdex/compare/05eec9ee0977dbaa577c007c96f1841688cb8a6f..7c46d81a71b020e6820091992722ad1e176207f6